datafusion/physical-plan/src/joins/cross_join.rs

```diff
@@ -46,16 +46,23 @@ use datafusion_physical_expr::equivalence::join_equivalence_properties;
 use async_trait::async_trait;
 use futures::{ready, Stream, StreamExt, TryStreamExt};
 
-/// Data of the left side
+/// Data of the left side that is buffered into memory
 type JoinLeftData = (RecordBatch, MemoryReservation);
 
-#[allow(rustdoc::private_intra_doc_links)]
-/// executes partitions in parallel and combines them into a set of
-/// partitions by combining all values from the left with all values on the right
+/// Cross Join Execution Plan
 ///
-/// Note that the `Clone` trait is not implemented for this struct due to the
-/// `left_fut` [`OnceAsync`], which is used to coordinate the loading of the
-/// left side with the processing in each output stream.
+/// This operator is used when there are no predicates between two tables and
+/// returns the Cartesian product of the two tables.
+///
+/// Buffers the left input into memory and then streams batches from each
+/// partition on the right input combining them with the buffered left input
+/// to generate the output.
+///
+/// # Clone / Shared State
+///
+/// Note this structure includes a [`OnceAsync`] that is used to coordinate the
+/// loading of the left side with the processing in each output stream.
+/// Therefore it can not be [`Clone`]
```

Comment:
Disregarding this PR -- I wonder is there any difference between cloning `OnceAsync` and `Arc<dyn ExecutionPlan>` fields?

It just seems that cloning both types is error prone for the cases when original operator is not dropped after being clones, and both clone and original trying to execute -- if so, and there is no significant difference between these types, maybe we should make `OnceAsync` cloneable, or forbid to clone almost all `PhysicalPlan`s?

Reply:
I think calling `ExecutionPlan::execute` results in a `Stream` 

In the join cases, the data would be shared across the clone, I think

For example, 

The first call to `HashJoinExec::execute`  (perhaps non obviously) updates the contents of `self.left_fut`  

https://github.com/apache/datafusion/blob/a32802c69cfb74ff5750700a8b72db87a44c55f0/datafusion/physical-plan/src/joins/hash_join.rs#L713-L712

And then subsequent invocations of HashJoinExec::exec read the (same) `self.left_fut` which is shared across streams

Reply:
Yes, this is correct.